itu-minitwit-cs/Program.cs

```diff
@@ -482,6 +482,60 @@ insert into user
     return Results.Content(render, "text/html; charset=utf-8");
 }
 
+app.MapMethods("/login", new[] { "GET", "POST" }, async (HttpRequest request, HttpContext context) =>
+{
+  // Logs the user in.
+
+  if (context.Items["user"] != null)
+  {
+    return Results.Redirect("/");
+  }
+
+  string error = null;
+  string username = null;
+
+  if (request.Method == "POST")
+  {
+    var form = await request.ReadFormAsync();
+    username = form["username"].ToString();
+    var password = form["password"].ToString();
+
+    var user = QueryDb((SqliteConnection)context.Items["db"], "SELECT * FROM user WHERE username = @p0",
+      new object[] { username }, one: true)?.FirstOrDefault();
+
+    if (user == null)
+    {
+      error = "Invalid username";
+    }
+    else if (!CheckPasswordHash(user["pw_hash"].ToString(), password))
+    {
+      error = "Invalid password";
+    }
+    else
+    {
+      context.Session.SetString("user_id", user["user_id"].ToString());
+      return Results.Redirect("/");
+    }
+  }
+
+  var data = new Dictionary<string, object>
+  {
+    ["error"] = error,
+    ["username"] = username
+  };
+
+  var finalRenderedHTML = sendToHtml(data, "login");
+  return Results.Content(finalRenderedHTML, "text/html; charset=utf-8");
+});
+
+bool CheckPasswordHash(string storedHash, string password)
+{
+  var pwHasher = new PasswordHasher<string>();
+
+  var verificationResult = pwHasher.VerifyHashedPassword((string)"abc", storedHash, password);
```

Comment:
I think the (string)"abc" here should be the username of the user whos password you are trying to hash
that is at least how the password is created in /register, and what the documentation for the method specifies:
[Password hasher documentation](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.passwordhasher-1.verifyhashedpassword?view=aspnetcore-9.0#microsoft-aspnetcore-identity-passwordhasher-1-verifyhashedpassword(-0-system-string-system-string))